File: classifier.php
<?php
/*
  Copyright (C) 2011  Glen Pitt-Pladdy

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA


  See: http://www.pitt-pladdy.com/blog/_20111229-214727_0000_Bayesian_Classifier_Classes_for_Perl_and_PHP/
*/

// based on: http://en.wikipedia.org/wiki/Bayesian_spam_filtering#Computing_the_probability_that_a_message_containing_a_given_word_is_spam

class classifier {
	private $dbh;
	private $words;
	private $driverclass;
	public $s = 3;
	public $unbiased = false;	// give even odds for all classes
	function classifier ( $dbh, $text ) {
		$this->dbh = $dbh;
		if ( strstr ( $dbh->getAttribute ( PDO::ATTR_DRIVER_NAME ), 'sqlite' ) === 0 ) {
			$this->driverclass = 'sqlite';
		} else {
			$this->driverclass = $dbh->getAttribute ( PDO::ATTR_DRIVER_NAME );
		}
		$this->words = preg_split ( '/\W+/', strtolower ( $text ) );
	}
	function teach ( $classification, $strength=1, $ordered=true ) {	// eg. 1=>HAM 2=>SPAM
		$prevwordid = NULL;
		$this->dbh->beginTransaction ();	// start transaction to avoid slow synchronous writes
		foreach ( $this->words as $word ) {
			if ( $word == '' ) { continue; }
			// put the word in the classification as needed
			// never delete words so this should be race safe
			$st = $this->dbh->prepare ( 'SELECT id FROM ClassifierWords WHERE Word = :word' );
			$st->execute ( array ( ':word'=>$word ) );
			$result = $st->fetch ();
			if ( isset ( $result['id'] ) ) {
				// got it
				$wordid = $result['id'];
			} else {
				// oops - missing word - add it
				$st = $this->dbh->prepare ( 'INSERT INTO ClassifierWords (Word) VALUES (:word)' );
				$st->execute ( array ( ':word'=>$word ) );
				$wordid = $this->dbh->lastInsertId ();
			}
			// SQLite has some limitations - work round them
			if ( $this->driverclass != 'sqlite' ) {
				// insert or update this word frequency
				$st = $this->dbh->prepare ( 'INSERT INTO ClassifierFrequency (Word,Class,Frequency) Values (:wordid,:class,:strength) ON DUPLICATE KEY UPDATE Frequency = Frequency + :strength' );
				$st->execute ( array ( ':wordid'=>$wordid, ':class'=>$classification, ':strength'=>$strength ) );
				// insert or update this word order frequency
				if ( $ordered ) {
					$st = $this->dbh->prepare ( 'INSERT INTO ClassifierOrderFrequency (Word,PrevWord,Class,Frequency) VALUES (:wordid,:prevwordid,:class,:strength) ON DUPLICATE KEY UPDATE Frequency = Frequency + :strength' );
					$st->execute ( array ( ':wordid'=>$wordid, ':prevwordid'=>$prevwordid, ':class'=>$classification, ':strength'=>$strength ) );
				}
			} else {
				// long way rount for SQLite
				// insert or update this word frequency
				$st = $this->dbh->prepare ( 'INSERT OR IGNORE INTO ClassifierFrequency (Word,Class,Frequency) Values (:wordid,:class,0)' );
				$st->execute ( array ( ':wordid'=>$wordid, ':class'=>$classification ) );
				$st = $this->dbh->prepare ( 'UPDATE ClassifierFrequency SET Frequency = Frequency + :strength WHERE Word = :wordid AND Class = :class' );
				$st->execute ( array ( ':strength'=>$strength, ':wordid'=>$wordid, ':class'=>$classification ) );
				// insert or update this word order frequency
				if ( $ordered ) {
					$st = $this->dbh->prepare ( 'INSERT OR IGNORE INTO ClassifierOrderFrequency (Word,PrevWord,Class,Frequency) VALUES (:wordid,:prevwordid,:class,0)' );
					$st->execute ( array ( ':wordid'=>$wordid, ':prevwordid'=>$prevwordid, ':class'=>$classification ) );
					$st = $this->dbh->prepare ( 'UPDATE ClassifierOrderFrequency SET Frequency = Frequency + :strength WHERE Word = :wordid AND PrevWord = :prevwordid AND Class = :class' );
					$st->execute ( array ( ':strength'=>$strength, ':wordid'=>$wordid, ':prevwordid'=>$prevwordid, ':class'=>$classification ) );
				}
			}
			// set for next word
			$prevwordid = $wordid;
		}
		// count classes
		if ( $this->driverclass != 'sqlite' ) {
			$st = $this->dbh->prepare ( 'INSERT INTO ClassifierClassSamples (Class,Frequency) VALUES (:class,:strength) ON DUPLICATE KEY UPDATE Frequency = Frequency + :strength' );
			$st->execute ( array ( ':class'=>$classification, ':strength'=>$strength ) );
		} else {
			// long way rount for SQLite
			$st = $this->dbh->prepare ( 'INSERT OR IGNORE INTO ClassifierClassSamples (Class,Frequency) VALUES (:class,0)' );
			$st->execute ( array ( ':class'=>$classification ) );
			$st = $this->dbh->prepare ( 'UPDATE ClassifierClassSamples SET Frequency = Frequency + :strength WHERE Class = :class' );
			$st->execute ( array ( ':strength'=>$strength, ':class'=>$classification ) );
		}
		$this->dbh->commit ();	// finish transaction to avoid slow synchronous writes
	}
	function classify ( $classifications, $useorder=0 ) {	// $useorder = 0-1 for the factor of order information
		$qualityfactor = count ( $classifications );
		// we need to know how many samples of each class to level the instances
		$messages = array ();
		$total = 0;
		$bindclassifications = implode ( ',', array_fill ( 0, count($classifications), '?' ) );
		foreach ( $classifications as $class ) {
			$st = $this->dbh->prepare ( 'SELECT Frequency FROM ClassifierClassSamples WHERE Class = :class' );
			$st->execute ( array ( ':class'=>$class ) );
			$result = $st->fetch();
			if ( ! isset ( $result['Frequency'] ) ) {
				// never seen this class before - can't classify
				return false;
			}
			$messages[$class] = $result['Frequency'];
			$total += $result['Frequency'];
		}
		// work out overall probability of each class
		$overallprob = array ();
		foreach ( $classifications as $class ) {
			$overallprob[$class] = $messages[$class] / $total;
		}
		// on with classification
		$prevword = NULL;
		$prob = array ();
		$proborder = array ();
		foreach ( $classifications as $class ) {
			$prob[$class] = 1;
			$proborder[$class] = 1;
		}
		foreach ( $this->words as $word ) {
			if ( $word == '' ) { continue; }
			// word frequency
			$st = $this->dbh->prepare ( 'SELECT Class,Frequency FROM ClassifierFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = ?) AND Class IN ('.$bindclassifications.')' );
			$count = 1;
			$st->bindValue ( $count++, $word );
			foreach ( $classifications as $class ) { $st->bindValue ( $count++, $class ); }
			$st->execute ();
			$results = $st->fetchAll ( PDO::FETCH_GROUP );
			$wordfrequency = array ();
			$total = 0;
			foreach ( $classifications as $class ) {
				if ( isset ( $results[$class][0]['Frequency'] ) ) {
					$wordfrequency[$class] = $results[$class][0]['Frequency'] / $messages[$class];
					$total += $results[$class][0]['Frequency'];
				} else {
					// no occurences of this word
					$wordfrequency[$class] = 0;
				}
			}
			if ( $total > 0 ) {	// no point otherwise
				$divisor = 0;
				$topline = array();
				foreach ( $classifications as $class ) {
						// could divide this by $total to give the prob per class, but it cancels anyway
						$topline[$class] = $wordfrequency[$class];
						if ( ! $this->unbiased ) { $topline[$class] *= $overallprob[$class]; }
						$divisor += $topline[$class];
				}
				foreach ( $classifications as $class ) {
					$probability = $topline[$class] / $divisor;
					// correct for few occurances
					$probability = $this->s * $overallprob[$class] + $total * $probability;
					$probability /= $this->s + $total;
					// account for word quality
					$quality = abs ( $probability - 1/$qualityfactor ) * $qualityfactor;
					if ( $quality < 0.3 ) { continue; }
					// combine
					$prob[$class] *= $probability;
				}
				// word order frequency
				$count = 1;
				if ( $prevword === NULL ) {
					$st = $this->dbh->prepare ( 'SELECT Class,Frequency FROM ClassifierOrderFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = ?) AND PrevWord IS NULL AND Class IN ('.$bindclassifications.')' );
					$st->bindValue ( $count++, $word );
				} else {
					$st = $this->dbh->prepare ( 'SELECT Class,Frequency FROM ClassifierOrderFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = ?) AND PrevWord = (SELECT id FROM ClassifierWords WHERE Word = ?) AND Class IN ('.$bindclassifications.')' );
					$st->bindValue ( $count++, $word );
					$st->bindValue ( $count++, $prevword );
				}
				foreach ( $classifications as $class ) { $st->bindValue ( $count++, $class ); }
				$st->execute ();
				$results = $st->fetchAll ( PDO::FETCH_GROUP );
				$total = 0;
				foreach ( $classifications as $class ) {
					if ( isset ( $results[$class][0]['Frequency'] ) ) {
						$wordfrequency[$class] = $results[$class][0]['Frequency'] / $messages[$class];
						$total += $results[$class][0]['Frequency'];
					} else {
						// no occurences of this word
						$wordfrequency[$class] = 0;
					}
				}
				if ( $total > 0 ) {	// no point otherwise
					$divisor = 0;
					foreach ( $classifications as $class ) {
							$topline[$class] = $wordfrequency[$class];
							if ( ! $this->unbiased ) { $topline[$class] *= $overallprob[$class]; }
							$divisor += $topline[$class];
					}
					foreach ( $classifications as $class ) {
						$probability = $topline[$class] / $divisor;
						// correct for few occurances
						$probability = $this->s * $overallprob[$class] + $total * $probability;
						$probability /= $this->s + $total;
						// account for word quality
						$quality = abs ( $probability - 1/$qualityfactor ) * $qualityfactor;
						if ( $quality < 0.3 ) { continue; }
						// combine
						$proborder[$class] *= $probability;
					}
				}
			}
			// set for next word
			$prevword = $word;
		}
		// finish up
		$total = 0;
		$totalorder = 0;
		foreach ( $classifications as $class ) {
			$total += $prob[$class];
			$totalorder += $proborder[$class];
		}
		foreach ( $classifications as $class ) {
			$prob[$class] /= $total;
			$proborder[$class] /= $totalorder;
		}
#error_log ( "scores:".var_export( $prob, true ) );
#error_log ( "scoresorder:".var_export( $proborder, true ) );
		// combine
		if ( $useorder > 0 ) {
			$usefreq = 1 - $useorder;
			foreach ( $classifications as $class ) {
				$probfreqweighted = pow ( $prob[$class], $usefreq );
				$proborderweighted = pow ( $proborder[$class], $useorder );
				$probcombined = $probfreqweighted * $proborderweighted;
				$prob[$class] = $probcombined / ( $probcombined + pow ( 1 - $prob[$class], $usefreq ) * pow ( 1 - $proborder[$class], $useorder ) );
			}
		}
#error_log ( "scoresfinal:".var_export( $prob, true ) );
		return $prob;
	}
	// do degrading of existing data so that fresh data takes prescedence
	function degrade ( $factor ) {	// 0.9 would multiply everything by that
		$this->dbh->beginTransaction ();	// start transaction to avoid slow synchronous writes
		$st = $this->dbh->prepare ( 'UPDATE ClassifierFrequency SET Frequency = Frequency * :factor' );
		$st->execute ( array ( ':factor'=>$factor ) );
		$st = $this->dbh->prepare ( 'UPDATE ClassifierOrderFrequency SET Frequency = Frequency * :factor' );
		$st->execute ( array ( ':factor'=>$factor ) );
		$st = $this->dbh->prepare ( 'UPDATE ClassifierClassSamples SET Frequency = Frequency * :factor' );
		$st->execute ( array ( ':factor'=>$factor ) );
		$this->dbh->commit ();	// finish transaction to avoid slow synchronous writes
	}
	// calculate quality factor for words
	function updatequality ( $limit=NULL ) {
		// get classes
		$st = $this->dbh->prepare ( 'SELECT DISTINCT Class FROM ClassifierFrequency' );
		$st->execute ();
		$classifications = $st->fetchAll ( PDO::FETCH_COLUMN );
		$qualityfactor = count ( $classifications );
		// we need to know how many samples of each class to level the instances
		$messages = array ();
		$total = 0;
		$bindclassifications = implode ( ',', array_fill ( 0, count($classifications), '?' ) );
		foreach ( $classifications as $class ) {
			$st = $this->dbh->prepare ( 'SELECT Frequency FROM ClassifierClassSamples WHERE Class = :class' );
			$st->execute ( array ( ':class'=>$class ) );
			$result = $st->fetch();
			if ( ! isset ( $result['Frequency'] ) ) {
				// never seen this class before - can't classify
				return false;
			}
			$messages[$class] = $result['Frequency'];
			$total += $result['Frequency'];
		}
		// work out overall probability of each class
		$overallprob = array ();
		foreach ( $classifications as $class ) {
			$overallprob[$class] = $messages[$class] / $total;
		}
		// get words
		if ( $limit !== NULL ) {
			if ( is_integer ( $limit ) ) {
				$st = $this->dbh->prepare ( 'SELECT Word FROM ClassifierWords ORDER BY LastUpdated LIMIT '.$limit );
			} else {
				return false;
			}
		} else {
			$st = $this->dbh->prepare ( 'SELECT Word FROM ClassifierWords ORDER BY LastUpdated' );
		}
		$st->execute ();
		$words = $st->fetchAll ( PDO::FETCH_COLUMN );
		// process each word
		$this->dbh->beginTransaction ();	// start transaction to avoid slow synchronous writes
		foreach ( $words as $word ) {
			// word frequency
			$st = $this->dbh->prepare ( 'SELECT Class,Frequency FROM ClassifierFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = ?) AND Class IN ('.$bindclassifications.')' );
			$count = 1;
			$st->bindValue ( $count++, $word );
			foreach ( $classifications as $class ) { $st->bindValue ( $count++, $class ); }
			$st->execute ();
			$results = $st->fetchAll ( PDO::FETCH_GROUP );
			$wordfrequency = array ();
			$total = 0;
			foreach ( $classifications as $class ) {
				if ( isset ( $results[$class][0]['Frequency'] ) ) {
					$wordfrequency[$class] = $results[$class][0]['Frequency'] / $messages[$class];
					$total += $results[$class][0]['Frequency'];
				} else {
					// no occurences of this word
					$wordfrequency[$class] = 0;
				}
			}
			$divisor = 0;
			$topline = array();
			foreach ( $classifications as $class ) {
				$topline[$class] = $wordfrequency[$class];
				if ( ! $this->unbiased ) { $topline[$class] *= $overallprob[$class]; }
				$divisor += $topline[$class];
			}
			if ( $divisor > 0 ) {
				$maxquality = 0;
				foreach ( $classifications as $class ) {
					$probability = $topline[$class] / $divisor;
					// correct for few occurances
					$probability = $this->s * $overallprob[$class] + $total * $probability;
					$probability /= $this->s + $total;
					// account for word quality
					$quality = abs ( $probability - 1/$qualityfactor ) * $qualityfactor;
					if ( $quality > $maxquality ) { $maxquality = $quality; }
				}
				$st = $this->dbh->prepare ( 'UPDATE ClassifierWords SET Quality = :quality, LastUpdated = :time WHERE Word = :word' );
				$st->execute ( array ( ':quality'=>$maxquality, ':time'=>time(), ':word'=>$word ) );
			} else {
				error_log ( "WARNING - ".__FILE__." - no Frequency data for word \"$word\"" );
				$st = $this->dbh->prepare ( 'UPDATE ClassifierWords SET LastUpdated = :time WHERE Word = :word' );
				$st->execute ( array ( ':time'=>time(), ':word'=>$word ) );
			}
		}
		$this->dbh->commit ();	// finish transaction to avoid slow synchronous writes
	}
}





?>

File: classifier.py
#
#
# Copyright (C) 2015  Glen Pitt-Pladdy
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
#
# See: https://www.pitt-pladdy.com/blog/_20150707-214047_0100_Bayesian_Classifier_Classes_for_Python/
# Previously: https://www.pitt-pladdy.com/blog/_20111229-214727_0000_Bayesian_Classifier_Classes_for_Perl_and_PHP/
#

# based on: http://en.wikipedia.org/wiki/Bayesian_spam_filtering#Computing_the_probability_that_a_message_containing_a_given_word_is_spam

import re
import time

class classifier:
	def __init__ ( self, db, text ):
		self.db = db;
		self.db.isolation_level = "DEFERRED"
		self.words = [ s[:40] for s in re.split ( '\W+', text.lower() ) ]	# TODO limit lengths in other flavours TODO
		self.s = 3
		self.unbiased = 0	# give even odds for all clases
		if str(type(self.db)) == "<class 'MySQLdb.connections.Connection'>":
			self.dbtype = 'MySQL'
		elif str(type(self.db)) == "<type 'sqlite3.Connection'>":
			self.dbtype = 'sqlite3'
	def teach ( self, classification, strength=None, ordered=None ):	# eg. 1=>HAM 2=>SPAM
		if strength == None: strength = 1.0
		if ordered == None: ordered = True
		prevwordid = None
		dbcur = self.db.cursor()
		# start transaction to avoid slow synchronous writes implied by isolation_level (above)
		for word in self.words:
			if word == '': continue
			# put the word in the classification as needed
			# never delete words so this should be race safe
			if self.dbtype == 'MySQL':
				dbcur.execute ( 'SELECT id FROM ClassifierWords WHERE Word = %s', [ word ] )
			else:
				dbcur.execute ( 'SELECT id FROM ClassifierWords WHERE Word = ?', [ word ] )
			result = dbcur.fetchone()
#			my $wordid;
			if result != None:
				# got it
				wordid = result[0]
			else:
				# oops - missing word - add it
				if self.dbtype == 'MySQL':
					dbcur.execute ( 'INSERT INTO ClassifierWords (Word) VALUES (%s) ON DUPLICATE KEY UPDATE Word=Word', [ word ] )
				else:
					dbcur.execute ( 'INSERT INTO ClassifierWords (Word) VALUES (?)', [ word ] )
				wordid = dbcur.lastrowid
			# SQLite has some limitations - work round them
			if self.dbtype == 'MySQL':
				# insert or update this word frequency
				dbcur.execute ( 'INSERT INTO ClassifierFrequency (Word,Class,Frequency) Values (%s,%s,%s) ON DUPLICATE KEY UPDATE Frequency = Frequency + %s', [ wordid, classification, strength, strength ] )
				# insert or update this word order frequency
				if ordered:
					dbcur.execute ( 'INSERT INTO ClassifierOrderFrequency (Word,PrevWord,Class,Frequency) VALUES (%s,%s,%s,%s) ON DUPLICATE KEY UPDATE Frequency = Frequency + %s', [ wordid, prevwordid, classification, strength, strength ] )
			else:
				# long way rount for SQLite
				# insert or update this word frequency
				dbcur.execute ( 'INSERT OR IGNORE INTO ClassifierFrequency (Word,Class,Frequency) Values (?,?,0)', [ wordid, classification ] )
				dbcur.execute ( 'UPDATE ClassifierFrequency SET Frequency = Frequency + ? WHERE Word = ? AND Class = ?', [ strength, wordid, classification ] )
				# insert or update this word order frequency
				if ordered:
					dbcur.execute ( 'INSERT OR IGNORE INTO ClassifierOrderFrequency (Word,PrevWord,Class,Frequency) VALUES (?,?,?,0)', [ wordid, prevwordid, classification ] )
					dbcur.execute ( 'UPDATE ClassifierOrderFrequency SET Frequency = Frequency + ? WHERE Word = ? AND PrevWord = ? AND Class = ?', [ strength, wordid, prevwordid, classification ] )
			# set for next word
			prevwordid = wordid;
		if self.dbtype == 'MySQL':
			dbcur.execute ( 'INSERT INTO ClassifierClassSamples (Class,Frequency) VALUES (%s,%s) ON DUPLICATE KEY UPDATE Frequency = Frequency + %s', [ classification, strength, strength ] )
		else:
			# long way rount for SQLite
			dbcur.execute ( 'INSERT OR IGNORE INTO ClassifierClassSamples (Class,Frequency) VALUES (?,0)', [ classification ] );
			dbcur.execute ( 'UPDATE ClassifierClassSamples SET Frequency = Frequency + ? WHERE Class = ?', [ strength, classification ] );
		self.db.commit()	# finish transaction to avoid slow synchronous writes
	def classify ( self, classifications, useorder=None ):	# $useorder = 0.0-1.0 for the factor of order information
		if useorder == None: useorder = 0.0
		qualityfactor = len( classifications )
		# we need to know how many samples of each clas to level the instances
		messages = {}
		total = 0
		bindclassifications = ''
		dbcur = self.db.cursor()
		for clas in classifications:
			if self.dbtype == 'MySQL':
				dbcur.execute ( 'SELECT Frequency FROM ClassifierClassSamples WHERE Class = %s', [ clas ] )
			else:
				dbcur.execute ( 'SELECT Frequency FROM ClassifierClassSamples WHERE Class = ?', [ clas ] )
			result = dbcur.fetchone()
			if result == None:
				# never seen this clas before - can't clasify
				return None
			messages[clas] = result[0]
			if messages[clas] == 0.0: messages[clas] = 1e-12	# use a safe tiny value so the math works
			total += messages[clas]
			if bindclassifications != '': bindclassifications += ','
			if self.dbtype == 'MySQL':
				bindclassifications += '%s';
			else:
				bindclassifications += '?';
		# work out overall probability of each clas
		overallprob = {}
		for clas in classifications:
			overallprob[clas] = messages[clas] / total
		# on with classification
		prevword = None
		prob = {}
		proborder = {}
		for clas in classifications:
			prob[clas] = 1
			proborder[clas] = 1
		for word in self.words:
			if word == '': continue
			# word frequency
			parameters = [ word ]
			parameters.extend ( classifications )
			if self.dbtype == 'MySQL':
				dbcur.execute ( 'SELECT Class,Frequency FROM ClassifierFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = %s) AND Class IN ('+bindclassifications+')', parameters )
			else:
				dbcur.execute ( 'SELECT Class,Frequency FROM ClassifierFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = ?) AND Class IN ('+bindclassifications+')', parameters )
			results = dbcur.fetchall()
			wordfrequency = {}
			total = 0
			for result in results:
				wordfrequency[result[0]] = result[1] / messages[clas]
				total += result[1]
			for clas in classifications:
				if not clas in wordfrequency: wordfrequency[clas] = 0.0
			if total > 0:	# no point otherwise
				divisor = 0.0
				topline = {}
				for clas in classifications:
						# could divide this by $total to give the prob per clas, but it cancels anyway
						topline[clas] = wordfrequency[clas]
						if not self.unbiased: topline[clas] *= overallprob[clas]
						divisor += topline[clas]
				zerorisk = False	# flag if we have a risk of zeroing out
				for clas in classifications:
					probability = topline[clas] / divisor
					# correct for few occurances
					probability = self.s * overallprob[clas] + total * probability
					probability /= self.s + total
					# account for word quality
					quality = abs ( probability - 1.0/qualityfactor ) * qualityfactor
					if quality < 0.3: continue
					# combine
					prob[clas] *= probability
					# check risk of zeroing out
					if prob[clas] <= 1e-200: zerorisk = True
				# if needed normalise to avoid zeroing out (rounding)
				if zerorisk:
					ntotal = 0.0
					for clas in classifications:
						ntotal += prob[clas]
					# avoid divide by Zero - we have probably rounded if this happens
					if ntotal == 0: ntotal = 1.0
					# convert to normal probabilities
					for clas in classifications:
						prob[clas] /= ntotal
				# word order frequency
				count = 1
				parameters = [ word ]
				if prevword == None:
					parameters.extend ( classifications )
					if self.dbtype == 'MySQL':
						dbcur.execute ( 'SELECT Class,Frequency FROM ClassifierOrderFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = %s) AND PrevWord IS NULL AND Class IN ('+bindclassifications+')', parameters )
					else:
						dbcur.execute ( 'SELECT Class,Frequency FROM ClassifierOrderFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = ?) AND PrevWord IS NULL AND Class IN ('+bindclassifications+')', parameters )
				else:
					parameters.append ( prevword )
					parameters.extend ( classifications )
					if self.dbtype == 'MySQL':
						dbcur.execute ( 'SELECT Class,Frequency FROM ClassifierOrderFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = %s) AND PrevWord = (SELECT id FROM ClassifierWords WHERE Word = %s) AND Class IN ('+bindclassifications+')', parameters )
					else:
						dbcur.execute ( 'SELECT Class,Frequency FROM ClassifierOrderFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = ?) AND PrevWord = (SELECT id FROM ClassifierWords WHERE Word = ?) AND Class IN ('+bindclassifications+')', parameters )
				results = dbcur.fetchall()
				total = 0
				for result in results:
					wordfrequency[result[0]] = result[1] / messages[clas]
					total += result[1]
				for clas in classifications:
					if not clas in wordfrequency: wordfrequency[clas] = 0.0
				if total > 0:	# no point otherwise
					divisor = 0.0
					for clas in classifications:
							topline[clas] = wordfrequency[clas]
							if not self.unbiased: topline[clas] *= overallprob[clas]
							divisor += topline[clas]
					zerorisk = False	# flag if we have a risk of zeroing out
					for clas in classifications:
						probability = topline[clas] / divisor
						# correct for few occurances
						probability = self.s * overallprob[clas] + total * probability
						probability /= self.s + total
						# account for word quality
						quality = abs ( probability - 1.0/qualityfactor ) * qualityfactor
						if quality < 0.3: continue
						# combine
						proborder[clas] *= probability
						# check risk of zeroing out
						if proborder[clas] <= 1e-200: zerorisk = True
					# if needed normalise to avoid zeroing out (rounding)
					if zerorisk:
						ntotal = 0.0
						for clas in classifications:
							ntotal += prob[clas]
						# avoid divide by Zero - we have probably rounded if this happens
						if ntotal == 0: ntotal = 1.0
						# convert to normal probabilities
						for clas in classifications:
							prob[clas] /= ntotal
			# set for next word
			prevword = word
		# finish up
		total = 0.0
		totalorder = 0.0
		for clas in classifications:
			total += prob[clas]
			totalorder += proborder[clas]
		# avoid divide by Zero - we have probably rounded if this happens
		if total == 0: total = 1.0
		if totalorder == 0: totalorder = 1.0
		# convert to normal probabilities
		for clas in classifications:
			prob[clas] /= total
			proborder[clas] /= totalorder
#		for clas in classifications:
#			print STDERR "scores: $clas => $prob[$clas]\n";
#			print STDERR "scoresorder: $clas => $proborder[$clas]\n";
		# combine
		if useorder > 0:
			usefreq = 1.0 - useorder
			for clas in classifications:
				probfreqweighted = prob[clas]**usefreq
				proborderweighted = proborder[clas]**useorder
				probcombined = probfreqweighted * proborderweighted
				prob[clas] = probcombined / ( probcombined + ( 1.0 - prob[clas] )**usefreq * ( 1.0 - proborder[clas] )**useorder )
#		for clas in classifications:
#			print STDERR "finalscores: $clas => $prob[$clas]\n";
		return prob
	# do degrading of existing data so that fresh data takes prescedence
	def degrade ( self, factor ):	# 0.9 would multiply everything by that
		dbcur = self.db.cursor()
		if self.dbtype == 'MySQL':
			dbcur.execute ( 'UPDATE ClassifierFrequency SET Frequency = Frequency * %s', [ factor ] )
			dbcur.execute ( 'UPDATE ClassifierOrderFrequency SET Frequency = Frequency * %s', [ factor ] )
			dbcur.execute ( 'UPDATE ClassifierClassSamples SET Frequency = Frequency * %s', [ factor ] )
		else:
			dbcur.execute ( 'UPDATE ClassifierFrequency SET Frequency = Frequency * :factor', [ factor ] )
			dbcur.execute ( 'UPDATE ClassifierOrderFrequency SET Frequency = Frequency * :factor', [ factor ] )
			dbcur.execute ( 'UPDATE ClassifierClassSamples SET Frequency = Frequency * :factor', [ factor ] )
		self.db.commit()	# finish transaction to avoid slow synchronous writes
	# remove words below a certain frequency (unlikely to be of value)
	def cleanfrequency ( self, threshold ):
		dbcur = self.db.cursor()
		if self.dbtype == 'MySQL':
			dbcur.execute ( 'DELETE FROM ClassifierFrequency WHERE Frequency < %s', [ threshold ] )
			dbcur.execute ( 'DELETE FROM ClassifierOrderFrequency WHERE Frequency < %s', [ threshold ] )
		else:
			dbcur.execute ( 'DELETE FROM ClassifierFrequency WHERE Frequency < :threshold', [ threshold ] )
			dbcur.execute ( 'DELETE FROM ClassifierOrderFrequency WHERE Frequency < :threshold', [ threshold ] )
		dbcur.execute ( 'DELETE FROM ClassifierWords WHERE id NOT IN (SELECT Word FROM ClassifierFrequency) AND id NOT IN (SELECT Word FROM ClassifierOrderFrequency) AND id NOT IN (SELECT PrevWord FROM ClassifierOrderFrequency WHERE PrevWord IS NOT NULL)' )
		self.db.commit()	# finish transaction to avoid slow synchronous writes
	# remove words below a certain quality (unlikely to be of value) TODO this is tricky since they may just be new ones
	# calculate quality factor for words
	def updatequality ( self, limit=None ):	# how many words to process (oldest first)
		dbcur = self.db.cursor()
		# get clases
		dbcur.execute ( 'SELECT DISTINCT Class FROM ClassifierFrequency' );
		classifications = [ row[0] for row in dbcur.fetchall() ]
		qualityfactor = len( classifications )
		# we need to know how many samples of each clas to level the instances
		messages = {}
		total = 0.0
		bindclassifications = ''
		for clas in classifications:
			if self.dbtype == 'MySQL':
				dbcur.execute ( 'SELECT Frequency FROM ClassifierClassSamples WHERE Class = %s', [ clas ] );
			else:
				dbcur.execute ( 'SELECT Frequency FROM ClassifierClassSamples WHERE Class = ?', [ clas ] );
			result = dbcur.fetchone ()
			if result == None:
				# never seen this clas before - can't clasify
				return None
			messages[clas] = result[0]
			total += result[0]
			if bindclassifications != '': bindclassifications += ','
			if self.dbtype == 'MySQL':
				bindclassifications += '%s'
			else:
				bindclassifications += '?'
		# work out overall probability of each clas
		overallprob = {}
		for clas in classifications:
			overallprob[clas] = messages[clas] / total
		# get words
		if limit != None:
			if limit.isdigit():
				dbcur.execute ( 'SELECT Word FROM ClassifierWords ORDER BY LastUpdated LIMIT %d' % limit )
			else:
				return None
		else:
			dbcur.execute ( 'SELECT Word FROM ClassifierWords ORDER BY LastUpdated' );
		words = [ row[0] for row in dbcur.fetchall() ]
		# process each word
		for word in words:
			# word frequency
			parameters = [ word ]
			parameters.extend ( classifications )
			if self.dbtype == 'MySQL':
				dbcur.execute ( 'SELECT Class,Frequency FROM ClassifierFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = %s) AND Class IN ('+bindclassifications+')', parameters );
			else:
				dbcur.execute ( 'SELECT Class,Frequency FROM ClassifierFrequency WHERE Word = (SELECT id FROM ClassifierWords WHERE Word = ?) AND Class IN ('+bindclassifications+')', parameters );
			results = dbcur.fetchall()
			wordfrequency = {}
			total = 0.0
			for result in results:
				wordfrequency[result[0]] = result[1] / messages[result[0]]
				total += result[1]
			for clas in classifications:
				if not clas in wordfrequency:
					# no occurences of this word
					wordfrequency[clas] = 0.0
			divisor = 0.0
			topline = {}
			for clas in classifications:
				topline[clas] = wordfrequency[clas]
				if not self.unbiased: topline[clas] *= overallprob[clas]
				divisor += topline[clas]
			if divisor > 0.0:
				maxquality = 0.0
				for clas in classifications:
					probability = topline[clas] / divisor
					# correct for few occurances
					probability = self.s * overallprob[clas] + total * probability
					probability /= self.s + total
					# account for word quality
					quality = abs ( probability - 1.0/qualityfactor ) * qualityfactor
					if quality > maxquality: maxquality = quality
				if self.dbtype == 'MySQL':
					dbcur.execute ( 'UPDATE ClassifierWords SET Quality = %s, LastUpdated = %s WHERE Word = %s', [ maxquality, int(time.time()), word ] )
				else:
					dbcur.execute ( 'UPDATE ClassifierWords SET Quality = :quality, LastUpdated = :time WHERE Word = :word', [ maxquality, int(time.time()), word ] )
			else:
# TODO				print STDERR "WARNING - ".__FILE__." - no Frequency data for word \"$word\"\n";
				if self.dbtype == 'MySQL':
					dbcur.execute ( 'UPDATE ClassifierWords SET LastUpdated = %s WHERE Word = %s', [ int(time.time()), word ] )
				else:
					dbcur.execute ( 'UPDATE ClassifierWords SET LastUpdated = :time WHERE Word = :word', [ int(time.time()), word ] )
		self.db.commit()	# finish transaction to avoid slow synchronous writes


